src/tls/cert.h

```diff
@@ -70,9 +70,10 @@ namespace tls
         peer_ca->use(ssl_ctx);
       }
 
-      if (auth != auth_default)
+      if (auth_required)
       {
-        SSL_CTX_set_verify(ssl_ctx, authmode(auth), NULL);
+        SSL_CTX_set_verify(
```

Comment:
This should be applied to `ssl` as well as it inherited the verification options from `ssl_ctx` when it was created, but hasn't been updated since. This leads to certification validation errors because of host name mismatches. I believe this was also a hidden problem with mbedTLS on the server side, where `mbedtls_ssl_conf_authmode` was not called when `Auth::auth_default` was enabled.

Reply:
There was an issue with adding that to the SSL object, I can't remember what it was. It could have been just that I was using the wrong flags, or a conflict in using the old `Auth` structure, but it'd be good to know what was wrong and how to do it right. It can be for a different path, if it's too convoluted.

Reply:
The `ssl` object is created just after `ssl_ctx`, with the default options and it won't be updated with the new settings via `SSL_CTX_set_verify`, so either we need to create it later, or duplicate the setting here (actually, just the `ssl` object would fine too I think). There's a problem with the peer host name not matching the name/SAN in the certificate, but I think that's just another matter of setting the right options.

Reply:
This is another thing we need to cleanup, to only create the SSL object at the last moment, so we only configure the context, not the session. But it was done that way to emulate mbedTLS.

Reply:
Yeah, I think so too. Alternatively, don't configure `ssl_ctx`, but consistently always `ssl`.

Reply:
In our use case, it doesn't matter because we always create both on a new connection. The easy solution is to ignore CTX and configure only SSL. This would still keep the format as we have today.

But we can also have a more persistent CTX and only create a new SSL for each connection. In that case, we'd need to split `Context` from `Connection` and only have `send` and `recv` on the latter.

I don't know what's the overhead of setting up, but I'd guess creating a new SSL from an existing (and already configured) CTX would be almost a copy construction.

The downside is that, if we have different types of connections (with and without verification, different certificates), then the persistent CTX needs to be a list of them, not just one, and we'd need a way to choose between them.

Reply:
Moved to  #3461 so I can resolve the conversation here.